src/java/org/apache/cassandra/replication/PendingLocalTransfer.java

```diff
@@ -0,0 +1,207 @@
+     * missed the TransferActivation. Transfers should not be pending for very long, and should be protected by
+     * internode integrity checks provided by TLS.
+     *
+     * TODO: Clear out the row cache and counter cache, like {@link CassandraStreamReceiver#finished}.
+     * TODO: Don't add to the live set if coordinator and not an owner for the range
+     */
+    public void activate(TransferActivation activation)
+    {
+        if (activated)
+            return;
+
+        Preconditions.checkState(isFullReplica());
+
+        logger.info("{} Activating transfer {}, {} ms since pending", logPrefix(), this, currentTimeMillis() - createdAt);
+        ColumnFamilyStore cfs = ColumnFamilyStore.getIfExists(tableId);
+        Preconditions.checkNotNull(cfs);
+        Preconditions.checkState(!sstables.isEmpty());
+
+        if (activation.dryRun)
+        {
+            logger.info("{} Not adding SSTables to live set for dryRun {}", logPrefix(), activation);
+            return;
+        }
+
+        // Ensure no lingering mutation IDs, only activation IDs
+        for (SSTableReader sstable : sstables)
+        {
+            Preconditions.checkState(sstable.getCoordinatorLogOffsets().isEmpty());
+
+            // Modify SSTables metadata to durably set transfer ID before importing
+            ImmutableCoordinatorLogOffsets logOffsets = new ImmutableCoordinatorLogOffsets.Builder()
+                                                  .addTransfer(activation.activationId)
+                                                  .build();
+            try
+            {
+                sstable.mutateCoordinatorLogOffsetsAndReload(logOffsets);
+            }
+            catch (IOException e)
+            {
+                throw new RuntimeException(e);
+            }
+
+            Preconditions.checkState(sstable.getCoordinatorLogOffsets().isEmpty());
+            Preconditions.checkState(!sstable.getCoordinatorLogOffsets().transfers().isEmpty());
+        }
+
+        File dst = cfs.getDirectories().getDirectoryForNewSSTables();
+
+        // Retain the original SSTables in pending/ dir on the coordinator, so future streams can get the originals and
+        // we don't need to isolate activated SSTables during compaction
+        boolean isCoordinator = activation.activationId.hostId == ClusterMetadata.current().myNodeId().id();
+        logger.debug("{} {} pending SSTables for activation to {}", isCoordinator ? "Copying" : "Moving", logPrefix(), dst);
+
+        dst.createFileIfNotExists();
+        for (SSTableReader sstable : sstables)
+        {
+            SSTableReader moved = SSTableReader.moveAndOpenSSTable(cfs, sstable.descriptor, cfs.getUniqueDescriptorFor(sstable.descriptor, dst), sstable.getComponents(), isCoordinator);
+            cfs.getTracker().addSSTablesTracked(Collections.singleton(moved));
+        }
+
+        activated = true;
```

Comment:
obviously we don't have durability implemented for this yet, but we should definitely have tests validating the restart safety of partially completed activations when we do